src/fixes/unordered_key.rs

```diff
@@ -0,0 +1,266 @@
+use super::Fix;
+use crate::common::*;
+
+pub(crate) struct UnorderedKeyFixer<'a> {
+    name: &'a str,
+}
+
+impl Default for UnorderedKeyFixer<'_> {
+    fn default() -> Self {
+        Self {
+            name: "UnorderedKey",
+        }
+    }
+}
+
+// When we sort the keys, we handle a significant line (with key) with all previous comments
+// as a whole. E. g.:
+// ```
+// B=C
+// # Comment
+// A=B
+// ```
+// will be fixed to:
+// ```
+// # Comment
+// A=B
+// B=C
+// ```
+// We check the order separately in each group of entries (groups are separated by blank lines).
+impl Fix for UnorderedKeyFixer<'_> {
+    fn name(&self) -> &str {
+        self.name
+    }
+
+    fn fix_warnings(
+        &self,
+        warnings: Vec<&mut Warning>,
+        lines: &mut Vec<LineEntry>,
+    ) -> Option<usize> {
+        // At first we find all groups
+
+        let mut sorted_part_started = false;
+        let mut start = None;
+        let mut end = None;
+
+        let mut sorted_groups = Vec::with_capacity(lines.len());
+
+        for (i, line) in lines.iter().enumerate() {
```

Comment:
I think that it is worth changing the approach here, the abundance of `Option` variables complicates the logic and requires unwanted `expect()`:thinking:

Maybe something like this? (quickly written code, may contain an error, just an attempt to convey my idea)
```
        let mut start_part_index = 0;

        for i in 0..lines.len()  {
            match lines.get(i) {
                Some(el) if el.is_empty() => {
                    if start_part_index < i {
                        let part = &mut lines[start_part_index..i];
                        sort_part(part);
                    }
                    start_part_index = i + 1;
                }
                _ => {}
            }
        }
```

This should be consistent with my other comments :slightly_smiling_face:

Reply:
I didn't quite get the idea (

We don't need `match` in this code snippet, we can use `[]`.

And it doesn't cover all cases (e.g. with comments).

There are 3 variables (`sorted_part_started`, `start`, `end`) and it seems that we need all of them.
A comment may or may not be the start of a sorted group:
```

# This is not
# a sorted group.
# No keys, just comments

```

Also `i` is not always the end of a sorted group:
```

A=B
C=D
# The end of a sorted group is
# the line `C=D`,
# not this one

```

Reply:
In fact, 3 variables seem redundant to me. I apologize that the code snippet was not clear.

Suppose in your version, the _start index_ is always shifted by 1 after finding an empty string (in next iteration). We may not use `Option` here.

In the case of `sorted_part_started`, I see a clear correlation with the `end`.

How do you like this option? He must pass all tests.

```
        let mut start = 0;
        let mut end = None;

        for i in 0..lines.len() {
            let line = &lines[i];
            let is_last_line = i == lines.len() - 1;

            if line.is_empty() || is_last_line {
                if let Some(end_index) = end {
                    Self::sort_part(
                        &mut lines[start..end_index],
                    );
                    end = None;
                }
                start = i + 1;
            } else if !line.is_comment() {
                end = Some(i + 1);
            }
        }
```

Perhaps this can be done even more expressively. But this is still the main idea.

Also, this option saves us from the need to check the last group outside the loop.

Reply:
I also noticed that we seem to be missing a test for something where the last line is not empty, but here we will get two `Warning`s (`UnorderedKey` and `EndingBlankLine`)

For example:
```dotenv
       # I comment\n\
        I=123\n\
        \n\
        B1=1\n\
        A1=1",
    );
```
Maybe add an integration test?


Therefore, to check the last line fully correct, we can slightly change the end condition:
```
        let mut start = 0;
        let mut end = None;

        for i in 0..lines.len() {
            let line = &lines[i];
            let is_last_line = i == lines.len() - 1;

            if !line.is_empty_or_comment(){
                end = Some(i + 1);
            }

            if line.is_empty() || is_last_line {
                if let Some(end_index) = end {
                    Self::sort_part(
                        &mut lines[start..end_index],
                    );
                    end = None;
                }
                start = i + 1;
            }
        }
```

Reply:
This solution is simpler, thank you.
I have used it (I have rename `start` to `start_index`) and have added the test (unit kind, I think this is OK).